controllers/hvpa_metrics.go

```diff
@@ -255,6 +264,192 @@ func addVPARecommendations(gv *prometheus.GaugeVec, vpaStatus *vpa_api.VerticalP
 	}
 }
 
+var allContainerResourceNames = [...]v1.ResourceName{
+	v1.ResourceCPU,
+	v1.ResourceMemory,
+}
+
+func addNaNVPARecommendations(gv *prometheus.GaugeVec, containers []string, baseLabelsFn func() prometheus.Labels) {
+	if gv == nil || len(containers) <= 0 {
+		return
+	}
+
+	addReco := func(container, recommendation string) {
+		for _, resource := range allContainerResourceNames {
+			l := baseLabelsFn()
+			l[labelContainer] = container
+			l[labelRecommendation] = recommendation
+			l[labelResource] = string(resource)
+
+			gv.With(l).Set(math.NaN())
+		}
+	}
+
+	for _, c := range containers {
+		for _, reco := range []string{recoTarget, recoLowerBound, recoUpperBound, recoUncappedTarget} {
+			addReco(c, reco)
+		}
+	}
+}
+
+func deleteVPARecommendations(gv *prometheus.GaugeVec, containers []string, baseLabelsFn func() prometheus.Labels) {
+	if gv == nil || len(containers) <= 0 {
+		return
+	}
+
+	deleteReco := func(container, recommendation string) {
+		for _, resource := range allContainerResourceNames {
+			l := baseLabelsFn()
+			l[labelContainer] = container
+			l[labelRecommendation] = recommendation
+			l[labelResource] = string(resource)
+
+			gv.Delete(l)
+		}
+	}
+
+	for _, c := range containers {
+		for _, reco := range []string{recoTarget, recoLowerBound, recoUpperBound, recoUncappedTarget} {
+			deleteReco(c, reco)
+		}
+	}
+}
+
+func (r *HvpaReconciler) getContainers(target runtime.Object) ([]string, error) {
+	var template v1.PodTemplateSpec
+	switch t := target.(type) {
+	case *appsv1.Deployment:
+		template = t.Spec.Template
+	case *appsv1.StatefulSet:
+		template = t.Spec.Template
+	case *appsv1.ReplicaSet:
+		template = t.Spec.Template
+	case *appsv1.DaemonSet:
+		template = t.Spec.Template
+	case *v1.ReplicationController:
+		template = *t.Spec.Template
+	default:
+		return nil, fmt.Errorf("Unhandled target type %T", t)
+	}
+
+	n := len(template.Spec.Containers)
+	containers := make([]string, n, n)
+	for i := range template.Spec.Containers {
+		containers[i] = template.Spec.Containers[i].Name
+	}
+
+	return containers, nil
+}
+
+func (r *HvpaReconciler) initScalingMetrics(hvpa *hvpav1alpha1.Hvpa, target runtime.Object) error {
+	if hvpa == nil || hvpa.Spec.TargetRef == nil {
+		log.V(3).Info("Invalid HVPA resource", "hvpa", hvpa)
+		return nil
+	}
+
+	var m = r.metrics
+
+	if m.specReplicas != nil && hvpa.Spec.Replicas != nil {
+		var replicas float64
+		if hvpa.Spec.Replicas != nil {
+			replicas = float64(*hvpa.Spec.Replicas)
+		}
```

Comment:
> This block can be replaced with:
`replicas := float64(*hvpa.Spec.Replicas)`
OR
`m.specReplicas.With(basicDetailedLabels(hvpa)).Set(float64(*hvpa.Spec.Replicas))`

Not really. My paranoia forces me to retain the nil check 😉

Reply:
> For now, until we fully implement scale subresource, should we set this field to desiredReplicas if the scaling succeeded, and to currentReplicas in case the scaling failed/blocked?

I think documenting it might be enough for now.

Reply:
> Not really. My paranoia forces me to retain the nil check 😉

It isn't required, since this is already checked 1 line above:
`if m.specReplicas != nil && hvpa.Spec.Replicas != nil {`

As is done in `updateScalingMetrics()`

Reply:
Ah. Missed that. Will do.